velox/vector/VectorPool.h

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "velox/vector/FlatVector.h"
+
+namespace facebook::velox {
+
+/// A thread-level cache of pre-allocated flat vectors of different types.
+/// Keeps up to 10 recyclable vectors of each of primitive types. A vector is
+/// recyclable if it is flat and recursively singly-referenced.
+class VectorPool {
+ public:
+  explicit VectorPool(memory::MemoryPool* pool) : pool_{pool} {}
+
+  VectorPtr get(const TypePtr& type, vector_size_t size);
```

Comment:
nit: shall we also comment say get will allocate a new vector if the typed cache pool is empty?